boa/src/environment/declarative_environment_record.rs

```diff
@@ -34,44 +34,51 @@ pub struct DeclarativeEnvironmentRecordBinding {
 /// declarations contained within its scope.
 #[derive(Debug, Trace, Finalize, Clone)]
 pub struct DeclarativeEnvironmentRecord {
-    pub env_rec: FxHashMap<Box<str>, DeclarativeEnvironmentRecordBinding>,
+    pub env_rec: GcCell<FxHashMap<Box<str>, DeclarativeEnvironmentRecordBinding>>,
     pub outer_env: Option<Environment>,
 }
 
 impl DeclarativeEnvironmentRecord {
+    pub fn new_declarative_environment_record(
+        env: Option<Environment>,
+    ) -> DeclarativeEnvironmentRecord {
+        DeclarativeEnvironmentRecord {
+            env_rec: GcCell::new(FxHashMap::default()),
+            outer_env: env,
+        }
+    }
+
     #[allow(clippy::new_ret_no_self)]
     pub fn new(env: Option<Environment>) -> Environment {
```

Comment:
I went with what required a smaller refactor, since those `new_`*something*`_environment_record` methods are not used that often. `new_declarative_environment_record` is never used outside of other environment records at all and `new_global_environment_record` is only used in `realm.rs`.

Reply:
Okay, I changed `new` methods to return `Self` and made a couple of adjustments to other parts of the code to deal with that. I managed to get rid of a couple of `borrow()` checks so that's nice.